napari/utils/colormaps/colormap.py

```diff
@@ -294,47 +316,77 @@ def _cast_labels_to_minimum_dtype_auto(
     np.ndarray
         Casted labels data.
     """
+    data = _convert_small_ints_to_unsigned(data)
+
+    if data.itemsize <= 2:
+        return data
+
+    num_colors = len(colormap.colors) - 1
+
     dtype = minimum_dtype_for_labels(num_colors + 1)
 
-    return _modulo_plus_one(data, num_colors, dtype, background_value)
+    if colormap.use_selection:
+        return (data == colormap.selection).astype(dtype) * (
+            _zero_preserving_modulo_naive(
+                np.array([colormap.selection]), num_colors, dtype
+            )
+        )
```

Comment:
I would split this out as well, it's hard to follow the logic with all these black indentations.